crates/wit-parser/src/decoding.rs

```diff
@@ -273,22 +410,128 @@ impl ComponentInfo {
             }
         }
 
-        let pkg = if let Some(name) = pkg_name {
-            Package {
+        let mut resolve = if let Some(name) = pkg_name {
+            let pkg = Package {
                 name,
                 docs: Docs::default(),
-                interfaces,
+                interfaces: interfaces.clone(),
                 worlds,
+            };
+            let (mut resolve, package) = decoder.finish(pkg);
+            implicit = Some(package);
+            // For now this is a sufficient condition to know that we're working with
+            // an implicit package declaration.  This will need to be reworked when
+            // mixed package declarations are supported
+            // if let Some(package) = implicit {
+            if let Some(package_metadata) = &self.package_metadata {
+                package_metadata.inject(&mut resolve, package)?;
             }
+            resolve
         } else {
-            bail!("no exported component type found");
+            Resolve::new()
         };
 
-        let (mut resolve, package) = decoder.finish(pkg);
-        if let Some(package_metadata) = &self.package_metadata {
-            package_metadata.inject(&mut resolve, package)?;
+        for explicit in &self.explicit {
```

Comment:
I think this is a copy/paste of the logic above? Can you please use helpers for significant chunks of code like this rather than copy/pasting and adapting?